Migrated the Edraak University ID feature with tests and code quality fixes

OmarIthawi opened the pull request:
### Description

TASK: [Migrate edraak_university module, with a feature flag.](https://app.asana.com/0/256944342521670/282932563583723)


### Docs

 - [Student Identity Mapping - User stories](https://docs.google.com/document/d/1pwvmL38bgiIyYJtnslgQG_Yz43HV0VxfxXgmJmuqDdQ/edit)


### Related Pull Requests

  - [x] Edraak university ID: Student view and admin panel https://github.com/Edraak/edx-platform/pull/233
- [x] Edraak University ID: Export ID information in the grade report https://github.com/Edraak/edx-platform/pull/240
 - [x] Edraak University ID: Instructor management view https://github.com/Edraak/edx-platform/pull/241
 - [x] Edraak university fixes: https://github.com/Edraak/edx-platform/pull/256

OmarIthawi commented:
@Salomari1987 The case is now left for the jury and the judge.

OmarIthawi commented:
Tears of joy 😭 my friend @Salomari1987 this has passed the CI tests.

Salomari1987 commented on lms/envs/common.py:
```diff
@@ -1427,6 +1427,9 @@
         'output_filename': 'css/lms-style-course-vendor.css',
     },
+    'style-edraak-university': {
+        'output_filename': 'css/lms-style-course-vendor.css',
```
source?

OmarIthawi replied:
😬 Not sure. Looks like a bug. Good catch @Salomari1987.

Salomari1987 commented on lms/djangoapps/edraak_university/tests/test_views.py:
```diff
@@ -0,0 +1,284 @@
+        soup = BeautifulSoup(res.content)
+        self.assertFalse(soup.select('ul.errorlist'), msg='Initial form should not contain errors')
+
+    @patch('edraak_university.views.update_account_settings')
```
Where is this from? And isnt the account_settings field to be removed? Refer to wiki

OmarIthawi replied:
This is a patch for the `update_account_settings` API. Due to the way  `mock.patch` works, it is always reliable to patch the local imported function (`edraak_university.views.update_account_settings`) instead of patching the original one (`user_api.accounts.api.update_account_settings`).

I'm not sure what do you mean by the `account_settings` field.

Salomari1987 commented:
@OmarIthawi CR looks good, left a comment for you, I will deploy it locally to do a quick QA

Salomari1987 commented:
@OmarIthawi I tested the app locally, and it behaves as suspected. It needs improvement in terms of design. It would also be great to add a feature whereby a staff member of the course can add a regex to define the rules for a valid university ID, this will reduce overhead. No?

OmarIthawi commented:
> @OmarIthawi I tested the app locally, and it behaves as suspected. It needs improvement in terms of design. It would also be great to add a feature whereby a staff member of the course can add a regex to define the rules for a valid university ID, this will reduce overhead. No?

There are number of planned features including the per-university validation that you've suggested. Currently the plan is to move the code as-is with tests and quality fixes. Future improvements will be done by @DQuraini.

Regarding the design. I think we're not planning to improve on, unless a client demands it.

OmarIthawi commented:
@Salomari1987 ready for another review:

 - Moved the settings to Edraak sections in the `/envs/` files.
 - Removed an incorrect css pipeline entry.

OmarIthawi commented:
Ready to merge @Salomari1987

Salomari1987 commented on lms/djangoapps/edraak_university/helpers.py:
```diff
@@ -0,0 +1,80 @@
+
+    if is_feature_enabled():
+        if course.enable_university_id:
+            return True
```
why not: is_feature_enabled() and course.get('enable_university_id', False) ? or just course.get(...)

OmarIthawi replied:
I doubt that `course.get(...)` can even work in this context. However I'll do the `and` part.

Salomari1987 replied:
this would raise KeyError exception if feature is not enabled

OmarIthawi replied:
It won't. The course object is not a dictionary. The field always exists regardless of the feature.

Salomari1987 commented on lms/djangoapps/edraak_university/migrations/__init__.py:
```diff
@@ -0,0 +1 @@
+# TODO: Link the UniversityID to the CourseEnrollment object to limit it for enrolled users only.
```
oops

OmarIthawi replied:
It's a big oops, [there's a task for it](https://app.asana.com/0/172236337713853/247229010586490).

Salomari1987 replied:
still won't fix itself ;)

Salomari1987 replied:
the \n at the end

Salomari1987 commented on lms/djangoapps/edraak_university/mixins.py:
```diff
@@ -0,0 +1,47 @@
+        Add the course to the context data.
+        """
+        data = super(CourseContextMixin, self).get_context_data(**kwargs)
+        data['course'] = self.get_course()
```
I would use .update(key=value)

OmarIthawi replied:
Overengineering?

Salomari1987 replied:
its faster

OmarIthawi replied:
I highly doubt this! Regardless, chopping few microseconds usually doesn't matter in the context of an HTTP request.

Salomari1987 replied:
👍

Salomari1987 commented on lms/djangoapps/edraak_university/forms.py:
```diff
@@ -0,0 +1,67 @@
+        },
+    )
+
+    section_number = forms.CharField(
```
Shouldnt this be cohort number, or something more understandable?

OmarIthawi replied:
edX already has a cohort system. And integration is planned but still not there. So eventually this should be done later.

Salomari1987 commented on lms/djangoapps/edraak_university/views.py:
```diff
@@ -0,0 +1,194 @@
+        return super(UniversityIDView, self).dispatch(*args, **kwargs)
+
+
+university_id = UniversityIDView.as_view()  # pylint: disable=invalid-name
```
Why?

OmarIthawi replied:
There used to be a very good comment about it. I'll get it back!

OmarIthawi commented:
@Salomari1987 ready for another go. I've addressed the issues.

Salomari1987 commented on lms/djangoapps/edraak_university/tests/tests.py:
```diff
@@ -0,0 +1,480 @@
+            with patch.dict(settings.FEATURES, EDRAAK_UNIVERSITY_APP=False):
+                self.assertEqual(
+                    first=tab.is_enabled(course, user),
+                    second=False,
```
do you utterly need this?

OmarIthawi replied:
Why not?

Salomari1987 replied:
by convention, ```first=``` and ```second=``` are not common in edX tests, so no need to use I think, not a big deal though, if you want to keep them then do.

OmarIthawi replied:
OK

OmarIthawi replied:
Done

Salomari1987 commented on lms/djangoapps/edraak_university/tests/tests.py:
```diff
@@ -0,0 +1,480 @@
+
+
+@ddt.ddt
+class TestHelpers(ModuleStoreTestCase):
```
HelpersTest

Salomari1987 commented on lms/djangoapps/edraak_university/helpers.py:
```diff
@@ -0,0 +1,79 @@
+    """
+    Correctly checks if the feature is enabled on a given course.
+    """
+    if not user and settings.ROOT_URLCONF == 'lms.urls':
```
Could you please elaborate, why this behaviour?

Salomari1987 commented on lms/djangoapps/edraak_university/tests/tests.py:
```diff
@@ -0,0 +1,480 @@
+
+    @patch.dict(settings.FEATURES, {'EDRAAK_UNIVERSITY_APP': True, 'EDRAAK_UNIVERSITY_CSV_EXPORT': True})
+    def test_is_csv_export_enabled_on_course_helper_enabled(self):
+        self.assertFalse(
```
lets do it like the test before, one line programming not good for huge projects

OmarIthawi replied:
Not sure that I get it

Salomari1987 replied:
what I mean, if this is done in multiple lines it would be clearer. Like 
``` 
@ddt.data((True, 'Export should be disabled when the feature flag is'), (False, 'Export should be enabled when all three features are enabled'))
@ddt.unpack
def test_name(self, enable_university_id, msg):

    is_csv_export_enabled_on_course = helpers.is_csv_export_enabled_on_course(Mock(enable_university_id=enable_university_id))
    assertEquals(is_csv_export_enabled_on_course, enable_university_id, msg=msg)
````

OmarIthawi replied:
Done

Salomari1987 commented on lms/djangoapps/edraak_university/tests/tests.py:
```diff
@@ -0,0 +1,480 @@
+            user=user,
+        )
+
+        self.assertIsNotNone(helpers.get_university_id(user, unicode(course.id)), 'Should not return an ID')
```
this is confusing, I guess the msg should be "Should return an ID", no?

OmarIthawi replied:
Done.

Salomari1987 replied:
Isn't the function testing whether get_university_id actually returns an id, so in case of failure we need to tell the developer that  **should return an ID** which is not what you are saying here

OmarIthawi replied:
Done

Salomari1987 commented on lms/djangoapps/edraak_university/tests/tests.py:
```diff
@@ -0,0 +1,480 @@
+
+        with patch('edraak_university.helpers.get_university_id', return_value=object()):
+            self.assertFalse(helpers.university_id_is_required(Mock(), Mock(enable_university_id=False)),
+                             'The course-specific feature flag is disabled, so the ID should not be required anyway')
```
how does that differ from the previous test Line 309

OmarIthawi replied:
Removed it. Useless test.

Salomari1987 commented on lms/djangoapps/edraak_university/tests/test_views.py:
```diff
@@ -0,0 +1,284 @@
+                         msg='The information was not valid, there should be no ID')
+
+        soup = BeautifulSoup(res.content)
+        self.assertTrue(len(soup.select('ul.errorlist')), msg='Should contain error messages')
```
Please add a test for the actually content, to see if the correct error message appears

OmarIthawi replied:
Done

Salomari1987 commented on lms/djangoapps/edraak_university/tests/test_views.py:
```diff
@@ -0,0 +1,284 @@
+        {'view_name': 'edraak_university_id_update', 'view_args': ['123']},
+        {'view_name': 'edraak_university_id_delete', 'view_args': ['123']},
+    )
+    def test_access_denied_for_students(self, view_name, view_args):
```
a better description is test_no_access_for_guests_and_students or something like that

OmarIthawi replied:
Done

Salomari1987 commented on lms/djangoapps/edraak_university/tests/test_views.py:
```diff
@@ -0,0 +1,284 @@
+        self.client.logout()
+        res_logged_out = self.client.get(course_url)
+
+        self.assertEquals(302, res_logged_out.status_code, '{} should redirect to login'.format(view_name))
```
Use assertRedirects(...

OmarIthawi replied:
Done

Salomari1987 commented on lms/djangoapps/edraak_university/tests/test_views.py:
```diff
@@ -0,0 +1,284 @@
+
+        self.assertNotContains(res, 'john_cooper_1895', msg_prefix='Should not contain the username')
+        self.assertContains(res, university_id.get_full_name(), msg_prefix='Should print the the full name')
+        self.assertContains(res, university_id.get_email(), msg_prefix='Should print the the email')
```
the the ... and again with full name

OmarIthawi replied:
Done

Salomari1987 commented on lms/djangoapps/edraak_university/tests/test_csv_export.py:
```diff
@@ -0,0 +1,147 @@
+        rows = self.create_rows()
+        row = self.get_row_as_dict(rows, index=1)
+
+        user = User.objects.get(email=self.EMAILS[0])
```
this could create problems in the future, if the EMAILS[0] was not as well an email with id, so I think using dictionary for emails would be more expressive.

OmarIthawi replied:
Good point. Done.

Salomari1987 commented on lms/djangoapps/edraak_university/tests/test_csv_export.py:
```diff
@@ -0,0 +1,147 @@
+        self.assertDictContainsSubset(university_id_data, row)
+
+    @patch.dict(settings.FEATURES, {'EDRAAK_UNIVERSITY_CSV_EXPORT': True})
+    def test_without_university_id(self):
```
Same here

OmarIthawi replied:
Done

OmarIthawi commented:
@Salomari1987 all notes have been addressed. Please review again.

Salomari1987 commented on lms/djangoapps/edraak_university/admin.py:
```diff
@@ -0,0 +1,59 @@
+            return None
+
+        return HTML(u'<a href="{url}">Profile</a>').format(
+            url=Text(reverse('admin:student_userprofile_change', args=[edraak_profile.pk])),
```
This will return NoReverseMatch, or 500 in production, please check ```student/admin.py``` for more details on new changes.

OmarIthawi replied:
Done

Salomari1987 commented on lms/djangoapps/edraak_university/urls.py:
```diff
@@ -0,0 +1,37 @@
+    '',
+    url(
+        r'^id/{}$'.format(settings.COURSE_ID_PATTERN),
+        'edraak_university.views.university_id',
```
Use namespace with ```^id$```

OmarIthawi replied:
?

OmarIthawi replied:
Done

OmarIthawi commented:
@Salomari1987 All comments have been addressed. Please review again.

Salomari1987 commented on lms/djangoapps/courseware/views/index.py:
```diff
@@ -388,7 +398,7 @@ def _create_courseware_context(self):
         courseware_context = {
             'csrf': csrf(self.request)['csrf_token'],
-            'COURSE_TITLE': self.course.display_name_with_default_escaped,
+            'COURSE_TITLE': self.course.display_name_with_default_escaped,  # safe-lint: disable=python-deprecated-display-name   pylint: disable=line-too-long
```
No need, the tests will pass regardless of the pylint disable

OmarIthawi replied:
@Salomari1987 This is a `safe-lint` not `pylint` as far as I know it doesn't. This is mainly an edX issue that I had to deal with because I've touched the file! The `safe-lint` tool is not that smart.

Salomari1987 commented on lms/djangoapps/courseware/views/index.py:
```diff
@@ -438,7 +448,7 @@ def _create_courseware_context(self):
 
             # section data
-            courseware_context['section_title'] = self.section.display_name_with_default_escaped
+            courseware_context['section_title'] = self.section.display_name_with_default_escaped  # safe-lint: disable=python-deprecated-display-name   pylint: disable=line-too-long
```
See commend above

Salomari1987 commented on lms/djangoapps/edraak_university/tests/tests.py:
```diff
@@ -0,0 +1,476 @@
+        {'urlconf': 'lms.urls', 'should_enable': False, 'msg': 'Should be disable when user=None in LMS'},
+        {'urlconf': 'cms.urls', 'should_enable': True, 'msg': 'Should be enabled when user=None in CMS'},
+    )
+    def test_enable_if_no_user_cms(self, urlconf, should_enable, msg):
```
Shouldnt the unpack come after the @ddt.data?

OmarIthawi replied:
@Salomari1987 Looks like not! It works both way.

OmarIthawi commented:
Ping @Salomari1987
